docs: blog: add Headlamp Kubeflow plugin blog (#56287)
kubernetes/website: content/en/blog/_posts/2026/introducing-headlamp-plugin-for-kubeflow/index.md

---
layout: blog
title: "Operating AI/ML workloads on Kubernetes: a Headlamp plugin for Kubeflow"

> date: 2026-05-26T10:00:00-08:00

Reviewer: ```suggestion
draft: true
```

Not sure why? See https://kubernetes.io/docs/contribute/blog/article-submission/

slug: introducing-headlamp-plugin-for-kubeflow
author: >
  [Alok Dangre](https://github.com/alokdangre) (independent)
---

> Kubernetes has quietly become the default platform for AI and machine learning. Whether you run notebook servers for data scientists, schedule distributed training jobs, tune hyperparameters, or orchestrate multi-step ML pipelines, those workloads increasingly land on a Kubernetes cluster. [Kubeflow](https://www.kubeflow.org/) is one of the most popular ways to assemble that stack, and it does so the Kubernetes-native way: every capability is exposed as a Custom Resource Definition (CRD).

Reviewer 2: Prefer Active voice : https://kubernetes.io/docs/contribute/style/style-guide/#use-active-voice

```suggestion
Kubernetes has quietly become the default platform for AI and machine learning. Whether you run notebook servers for data scientists, schedule distributed training jobs, tune hyperparameters, or orchestrate multi-step ML pipelines, those workloads increasingly land on a Kubernetes cluster. [Kubeflow](https://www.kubeflow.org/) is one of the most popular ways to assemble that stack, and it does so the Kubernetes-native way. It exposes every capability as a Custom Resource Definition (CRD).
```

That design is a gift to cluster operators, because it means ML workloads can be observed and managed with the same primitives as everything else in the cluster. But in practice the specialized ML dashboards that ship with these platforms hide the Kubernetes layer underneath. When a notebook is stuck or a training run fails, the operator is often left dropping back to `kubectl` to find out what actually happened at the Pod level.

This post introduces the **Headlamp Kubeflow plugin**, which closes that gap by surfacing Kubeflow's custom resources directly inside a general-purpose Kubernetes UI. It is a worked example of a pattern any CRD-heavy platform can follow: meet operators where they already work, and show them the cluster-level truth.

> Headlamp itself is an extensible Kubernetes web UI maintained under [Kubernetes SIG UI](https://github.com/kubernetes-sigs/headlamp) and licensed under Apache 2.0. It runs as a desktop app or in-cluster, and its plugin system lets anyone add first-class views for custom resources.

Reviewer 2: The link text says Kubernetes SIG UI but points to the Headlamp repository.

I would say add the headlamp link in the Headlamp first word and separate link for sig-ui
```suggestion
[Headlamp](https://github.com/kubernetes-sigs/headlamp) itself is an extensible Kubernetes web UI maintained under [Kubernetes SIG UI](https://github.com/kubernetes/community/blob/main/sig-ui/) and licensed under Apache 2.0. It runs as a desktop app or in-cluster, and its plugin system lets anyone add first-class views for custom resources.
```

| **Katib** | Automated hyperparameter tuning and neural architecture search (AutoML) | Experiment, Trial, Suggestion |
| **Training** | Runs distributed model training jobs (for example PyTorch or TensorFlow) on the cluster | TrainJob, TrainingRuntime, ClusterTrainingRuntime |
| **Spark** | Large-scale data processing with Apache Spark on Kubernetes | SparkApplication, ScheduledSparkApplication |

## What you can see

### Pod-level truth for notebooks

Every resource type has a detail view built for operators rather than end users. A notebook server, for instance, shows the actual Kubernetes Pod conditions with their `reason` and `message` fields, the full resource requests and limits including GPUs, volume mounts mapped to their backing storage type (PVC, ConfigMap, Secret, or EmptyDir), environment variables including Secret and ConfigMap references, and any sidecar containers or node tolerations. This is the information you would otherwise piece together from several `kubectl describe` calls.

### Hyperparameter tuning at a glance

> The Katib views expose the whole tuning workflow: the algorithm and search space, every trial with live status, and the current best trial with its metric values and parameter assignments. Early stopping configuration and the count of early-stopped trials are shown alongside, so you can see how a search is converging without leaving the cluster UI.

Reviewer 2: Use UpperCamelCase for API kinds, and Trial is a Katib CRD : https://kubernetes.io/docs/contribute/style/style-guide/#use-upper-camel-case-for-api-objects

```suggestion
The Katib views expose the whole tuning workflow: the algorithm and search space, every Trial with live status, and the current best Trial with its metric values and parameter assignments. Early stopping configuration and the count of early-stopped trials are shown alongside, so you can see how a search is converging without leaving the cluster UI.
```

### Pipeline state without the pipeline database

> The Pipelines views read straight from Kubernetes CRDs and do not touch the ML pipeline backend database. You can inspect pipeline state even when the pipeline service itself is unavailable. The views surface version history with a side-by-side YAML diff between versions, run state and duration, recurring schedules rendered in human-readable form, and the artifact storage roots discovered across recent runs.

Reviewer 2: ```suggestion
The Pipelines views read straight from Kubernetes CRDs and do not touch the ML pipeline backend database. You can inspect pipeline state even when the pipeline service itself is unavailable. The views surface version history with a side-by-side YAML diff between versions, run state and duration, human-readable recurring schedules, and the artifact storage roots discovered across recent runs.
```

### A topology map for ML resources

> The plugin contributes a [Headlamp map source](https://github.com/kubernetes-sigs/headlamp) that renders Kubeflow resources, including Notebooks, Profiles, Katib Experiments, Pipelines, and Spark Applications, as a graph. Ownership relationships between resources become edges, giving you a topology view of how the pieces fit together. Hovering over any of these resources elsewhere in Headlamp also brings up an inline summary, so the context travels with you.

Reviewer 2: The link on Headlamp map source currently points to the generic Headlamp repository root which does not explain what a map source is. Replace it with the map source page that covers this concept specifically:

```suggestion
The plugin contributes a [Headlamp map source](https://headlamp.dev/docs/latest/development/plugins/functionality/extending-the-map/) that renders Kubeflow resources, including Notebooks, Profiles, Katib Experiments, Pipelines, and Spark Applications, as a graph. Ownership relationships between resources become edges, giving you a topology view of how the pieces fit together. Hovering over any of these resources elsewhere in Headlamp also brings up an inline summary, so the context travels with you.
```

## Try it

> The plugin is available as an alpha release in the Headlamp repository. Installation and local-cluster setup instructions, including a lightweight CRD-only path for evaluation, live in the [plugin README](https://github.com/kubernetes-sigs/headlamp/tree/main/plugins/kubeflow). Because it works against any subset of installed components, you can point it at an existing Kubeflow cluster or spin up a throwaway cluster with only the CRDs to explore the UI.

Reviewer 2: Link show `404` error
```suggestion
The plugin is available as an alpha release in the Headlamp repository. Installation and local-cluster setup instructions, including a lightweight CRD-only path for evaluation, live in the [plugin README](https://github.com/headlamp-k8s/plugins/tree/main/kubeflow). Because it works against any subset of installed components, you can point it at an existing Kubeflow cluster or spin up a throwaway cluster with only the CRDs to explore the UI.
```
